V2 Wizard: Add snapshot method selection to the image builder wizard 2 (HMS-3796)

Andrewgdewar opened the pull request:
- Adds the following screens for snapshotting functionality.


 ![Screenshot 2024-04-17 at 3 28 08 PM](https://github.com/osbuild/image-builder-frontend/assets/38083295/3f823519-fffe-49b4-9415-f3844b0c45ac)
![Screenshot 2024-04-17 at 3 28 11 PM](https://github.com/osbuild/image-builder-frontend/assets/38083295/2bee9438-b1cd-4bdf-ade9-649f514100e3)
![Screenshot 2024-04-17 at 3 28 22 PM](https://github.com/osbuild/image-builder-frontend/assets/38083295/8d2b0605-947c-46f7-aa76-84e84457d384)
![Screenshot 2024-04-17 at 3 34 31 PM](https://github.com/osbuild/image-builder-frontend/assets/38083295/6c19b331-5e83-469a-91e3-c256bdb5d46e)
![Screenshot 2024-04-17 at 3 29 28 PM](https://github.com/osbuild/image-builder-frontend/assets/38083295/b5509b7e-0d4c-4518-8cb2-24b1c64ff6dc)


TODO:
- [ ]  Add test for snapshot functionality

lucasgarfield commented on src/Components/CreateImageWizardV2/steps/Snapshot/Snapshot.tsx:
```diff
@@ -0,0 +1,145 @@
+  const setSnapshotDate = (val: string) => dispatch(changeSnapshotDate(val));
+
+  // TODO: Remove query once snapshotting is no longer a "feature flag" i.e. when in Prod-stable
+  // Ensure to remove the Alert and ConditionalTooltip below.
+  const { data, isSuccess, isFetching, isError } = useListFeaturesQuery();
+  const isSnapshotDisabled =
+    isError ||
+    isFetching ||
```
When do we expect this to hit prod-stable? Should we consider putting this step behind an Unleash feature flag? We would like to get the V2 wizard into prod-preview next week with prod-stable to follow shortly thereafter.

Andrewgdewar replied:
In prod-stable the other toggle will simply be disabled for now, on hover of the disabled "Use a snapshot" Radio you will see "Snapshotting is not currently available". 

If this is acceptable until snapshotting is live in Prod-stable, then I'd say that you already have your feature flag handled.

lucasgarfield replied:
Would it be available in prod-preview? UX has developed some patterns for us where features are disabled in prod-stable but users are encouraged to check them out in prod-preview. We could try that if this is available in prod-preview (in a follow-up PR).

Andrewgdewar replied:
Okay I added some better logic to handle this, it is ugly, but temporary. 
Attempting to incorporate redux-like storage would make pinging the features api much more complex, I would like to avoid that for what should be a very simple change.

Andrewgdewar replied:
<img width="695" alt="Screenshot 2024-04-22 at 10 34 22 PM" src="https://github.com/osbuild/image-builder-frontend/assets/38083295/7fb75606-b117-4761-9da5-967326b2b973">

lucasgarfield commented on src/Components/CreateImageWizardV2/steps/Snapshot/Snapshot.tsx:
```diff
@@ -0,0 +1,145 @@
+            onChange={() => useLatest && setUseLatest(false)}
+          />
+        </ConditionalTooltip>
+        {isError && (
```
Would we also remove this when snapshotting goes into prod-stable?

Andrewgdewar replied:
Indeed

lucasgarfield reviewed (changes requested):
This is amazing @Andrewgdewar! I love all the little improvements you've made like converting `time.js` to `time.ts`. Thanks!

I have a couple of questions/nitpicks and two major requests:
1. We have two "modes" for the wizard. "Create" mode where we create a new blueprint, and "edit" mode where we updated a blueprint. In `requestMapper.tsx` you updated the `mapRequestToState` function which takes care of initializing the state in "edit" mode. We still need to make sure that we actually send the request correctly, and for that I think you'll need to add the `snapshot_date` field to the `getCustomizations` function.
2. Would you mind also adding a test that verifies the request is generated correctly when snapshotting is being used? You can follow the pattern in `Details.test.tsx` for instance. It should be easy to get passing assuming you get (1) fixed. :wink:

Andrewgdewar commented on src/Components/CreateImageWizardV2/CreateImageWizard.tsx:
```diff
@@ -139,6 +142,10 @@ const CreateImageWizard = ({ startStepIndex = 1 }: CreateImageWizardProps) => {
+  const snapshotDate = useAppSelector(selectSnapshotDate);
+  const useLatest = useAppSelector(selectUseLatest);
+
+  const snapshotStepRequiresChoice = !useLatest && !snapshotDate;
```
I'll take a look!

Andrewgdewar replied:
Hmmm, I don't think his functionality matches my own, we may want to look at something a bit more cohesive for both, as I think a better pattern is to stop the user from progressing should they change a required field.

The current implementation of your input "details" input is not triggered "onChange" but rather onBlur, which would allow the user to click away. 

My implementation prevents the movement to the next step altogether. I think a fusion of these two approaches would be ideal. 

An input validator that changes on the "onChange" handler, marks an error in the field as @ezr-ondrej  PR accomplishes, but also prevents movement to the next step. 

<img width="677" alt="Screenshot 2024-04-22 at 7 51 14 PM" src="https://github.com/osbuild/image-builder-frontend/assets/38083295/0a32ed11-ca06-4cd0-9f39-41234ebd7fc1">
<img width="616" alt="Screenshot 2024-04-22 at 7 52 03 PM" src="https://github.com/osbuild/image-builder-frontend/assets/38083295/3558b1ce-99d4-48fd-bc28-04570e99d834">

I also noticed a bug with the details step due to that functionality. 
If navigate to the details step during edit, I cannot "step" as the "onBlur" event hasn't been fired yet.:
<img width="593" alt="Screenshot 2024-04-22 at 7 53 39 PM" src="https://github.com/osbuild/image-builder-frontend/assets/38083295/9e8b4ff6-be0d-4987-b8ea-74d3d1e8df73">

ezr-ondrej replied:
My PR allows to do both, we just need to consume the validation state here and use it for `disableNext`.
I don't think it's perfect, but it allows for validation to be handled in a single place.
The way we do it currently requires the validation to be done here and on the component.

Nice find on the bug tho, that's bit tricky on my approach.
Might actually be best to validate directly in the Redux action :thinking:

Andrewgdewar replied:
I'll look into this, I think at the minimum we'd want to add a debounce to the input to allow for instant feedback on state.

ezr-ondrej replied:
I'd love to team up and come up with a better way to validate.
That said, I'd not block this PR on it.

Andrewgdewar commented on src/Components/CreateImageWizardV2/steps/Repositories/Repositories.tsx:
```diff
@@ -320,7 +320,11 @@ const Repositories = () => {
+        data.data
+          ?.filter(({ status }) => status === 'Valid')
+          .map((repo) => repo.url) || []
+      );
```
Indeed, and now that you mention it, it is an incomplete bug fix. I need to make sure the that selectAll button is greyed out if none of the repositories are valid as well...

Andrewgdewar replied:
I think I could easily rewrite the repositories piece to match our own new template wizard which is very similar. 
If you'd like me to maintain it I'd suggest doing that so the look/feel is identical between both wizards for that step.
Also because of the similarity in syntax between this flavour of Redux and react-query I think it would be an easy lift.

Andrewgdewar replied:
Feel free to throw-up a card and ping me on it if you'd like me to do that.

Andrewgdewar commented on src/Components/CreateImageWizardV2/steps/Snapshot/Snapshot.tsx:
```diff
@@ -0,0 +1,145 @@
+  const snapshotDate = useAppSelector(selectSnapshotDate);
+  const useLatest = useAppSelector(selectUseLatest);
+  const setUseLatest = (val: boolean) => dispatch(changeUseLatest(val));
+  const setSnapshotDate = (val: string) => dispatch(changeSnapshotDate(val));
```
Could do, I am visitor 😄

Andrewgdewar commented:
> 1. We have two "modes" for the wizard. "Create" mode where we create a new blueprint, and "edit" mode where we updated a blueprint. In `requestMapper.tsx` you updated the `mapRequestToState` function which takes care of initializing the state in "edit" mode. We still need to make sure that we actually send the request correctly, and for that I think you'll need to add the `snapshot_date` field to the `getCustomizations` function.

I'm not sure why I would need to add that field to the "getCustomizations" function, as I have added the field to the request.image_requests? 

Why would I store the date their when I presume that the date should be stored on the image?

Andrewgdewar commented:
Added empty state for when no repositories are selected.
Feedback on copy anyone?
<img width="801" alt="Screenshot 2024-04-22 at 11 13 51 PM" src="https://github.com/osbuild/image-builder-frontend/assets/38083295/ca8e31f2-6d06-49c1-bb50-cab1b2c4df9e">

lucasgarfield commented:
> @lucasgarfield
> 
> > 1. We have two "modes" for the wizard. "Create" mode where we create a new blueprint, and "edit" mode where we updated a blueprint. In `requestMapper.tsx` you updated the `mapRequestToState` function which takes care of initializing the state in "edit" mode. We still need to make sure that we actually send the request correctly, and for that I think you'll need to add the `snapshot_date` field to the `getCustomizations` function.
> 
> I'm not sure why I would need to add that field to the "getCustomizations" function, as I have added the field to the request.image_requests?
> 
> I might be missing something... 🤔
> 
> Why would I store the date there when I presume that the date should be stored on the image?

:facepalm: You're right, I got confused. I thought this got stored in the `customizations` but it is indeed in `image_requests`. Sorry for the confusion!

lucasgarfield reviewed (approved):
This is amazing @Andrewgdewar, the new tests look great. I squashed the commits so we have a clean history and fixed the merge conflicts, hope that's okay with you. :wink: This is ready to go in.
